### PR TITLE
Added Dynamic Search and Filter Button in Challenge List Page

### PR DESCRIPTION
**Pull request by wahid18-maqs**
# Description

This PR introduces two new features to the Challenge List page on the EvalAI platform:

1. Dynamic Search Function
- A dynamic search bar has been implemented to allow users to search through **Ongoing Challenges**, **Upcoming Challenges**, and **Past Challenges** within their respective tabs.
- As the user types, the challenge list will be filtered based on the search query, providing a faster and more efficient way to navigate through challenges.
 2. Filter Button
- A **Filter** button has been added that enables users to filter challenges based on specific criteria (e.g., organization name, tags, start date, and end date).
- This enhances the user experience by allowing for more precise filtering options.

# Motivation

The addition of the dynamic search function and filter button significantly improves the user experience when navigating the challenge list on EvalAI. With the increasing number of hosted challenges, it is essential to provide users with efficient ways to find relevant challenges. These features aim to enhance usability and accessibility, making EvalAI more user-friendly.

# Changes Made in Files
 `frontend/src/views/web/challenge-list.html`
- Added a **Search and Filter Bar** at the top.
- Implemented a **Search Bar** for real-time filtering of challenges.
- Added a **Filter Button** to toggle the visibility of a filter side panel.
- Introduced a toggleable **Filter Panel**:
  - Filters include **Organization**, **Tags**, and sorting options like **Start Date** and **End Date**.
  - Added **Apply** and **Reset** buttons inside the panel.
- Integrated filter pipe into ongoing, upcoming, and past challenge sections for dynamic search.
- Displayed a conditional message when no challenges match the applied search or filter criteria.

`frontend/src/js/controllers/challengeListCtrl.js`
- **Filtering & Sorting**:
  - Organization and tag-based filtering.
  - Sort by start/end date.
- **UI Responsiveness**:
  - Manual `$scope.$apply()` and `$timeout()` for safe asynchronous updates.
- **API Handling**:
  - Graceful fallback for missing data and Basic error messaging (`vm.errorMessage`).
- **Time Zone & Date Handling**:
  - Parsed `start_date` and `end_date` as Date objects, added `time_zone` and `gmt_zone` fields.
- **State Management**:
  - Centralized filter state setup, Consistent flags: `vm.hasAppliedFilter`, `vm.isFilterVisible`.
 

 `frontend/tests/controllers-test/challengeListCtrl.test.js`
- Added centralized API error handling with `handleApiError()`.
- Introduced `processChallenges()` to enrich challenge data (timezone, creator mapping, date objects).
- Improved pagination with recursive fetching using `vm.getAllResults()`.
- Enhanced filtering and sorting capabilities (organization, tags, start/end date).
- Improved code readability and maintainability through modular functions.
- Added scroll-based button visibility handling with `vm.scrollUp()`.
- Improved loader management during API calls.

 `frontend/src/css/modules/challenge.scss`
- Added styles for **Search and Filter Bar** at the top.
- Added styles for the **New Filter Panel**.
- Improved card styling and responsive design.
- Enhanced the "No Results" display.
- Added new utility classes.

# Screenshots
- Working of search function and filter button.

https://github.com/user-attachments/assets/61a262e1-c215-4e36-8cd4-8d61b30f3794

- Result of Unit Test 

seach-filter-test

### **Checklist**:
- [x] UI changes have been tested manually.
- [x] New functionality is working as expected.
- [x] No console errors or warnings.
- [x] Code is properly formatted and follows project conventions.

### CODE REVIEW

**Comment by wahid18-maqs** [comment]
Although no specific issue was created for this feature i have developed this to improve the usability and navigation on the on the challenge list page

**Comment by wahid18-maqs** [comment]
@RishabhJain2018 Sir please review it !!

File: frontend/src/js/controllers/challengeListCtrl.js [hunk 1]
```diff
@@ -17,97 +17,255 @@
...
+                        }
+
+                        resultsArray.push(challenge);
+                    });
 
-                    // check for the next page
                     if (data.next !== null) {
                         var url = data.next;
-                        var slicedUrl = url.substring(url.indexOf('challenges/challenge'), url.length);
+                        var slicedUrl = url.substring(url.indexOf('challenges/challenge'));
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeListCtrl.js:96]: Why are we changing this line?

File: frontend/src/js/controllers/challengeListCtrl.js [hunk 2]
```diff
@@ -17,97 +17,255 @@
...
                         parameters.url = slicedUrl;
-                        vm.getAllResults(parameters, resultsArray, typ);
+                        vm.getAllResults(parameters, resultsArray, type);
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeListCtrl.js:98]: Same as above!

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
Hi @wahid18-maqs , Thanks for the feature. Can you please match the `selected field` and `Show filters` color to match the website theme?
Screenshot 2025-04-27 at 11 22 45 PM

File: frontend/src/js/controllers/challengeListCtrl.js (hunk 1 above)
Comment (wahid18-maqs) [reply frontend/src/js/controllers/challengeListCtrl.js:96]: I have removed `url.length` since `substring(startIndex)` already slices to the end by default, making the code cleaner without changing functionality — but I'll add it back now.

File: frontend/src/js/controllers/challengeListCtrl.js (hunk 2 above)
Comment (wahid18-maqs) [reply frontend/src/js/controllers/challengeListCtrl.js:98]: Here I have renamed `typ` ➔ `type` to make the code more readable.  
There is no change in functionality, only an improvement in naming clarity.

**Comment by wahid18-maqs** [comment]
Matched the selected field and Show Filters color with the website theme, and re-added `url.length` in `frontend/src/js/controllers/challengeListCtrl.js`.


https://github.com/user-attachments/assets/403fec43-af2f-4809-b651-6e75a321634c

**Comment by RishabhJain2018** [comment]
Hi @wahid18-maqs , Thanks for the fixes. Can we please do the same formatting changes for this filter?
Screenshot 2025-04-28 at 11 46 49 AM

**Comment by wahid18-maqs** [comment]
Sure @RishabhJain2018 Sir

**Comment by wahid18-maqs** [comment]
changed format of filter panel 

https://github.com/user-attachments/assets/bc6789ec-f496-4ca6-9d6b-ec2343606c0c

**Comment by wahid18-maqs** [comment]
@RishabhJain2018 sir please review it !!

File: frontend/tests/controllers-test/challengeListCtrl.test.js [hunk 3]
```diff
@@ -1,368 +1,207 @@
...
-
-    beforeEach(inject(function (_$controller_, _$rootScope_, _utilities_,) {
-        $controller = _$controller_;
-        $rootScope = _$rootScope_;
-        utilities = _utilities_;
-
-        $scope = $rootScope.$new();
-        createController = function () {
-            return $controller('ChallengeListCtrl', {$scope: $scope});
+angular.module('evalai')
```
Comment (RishabhJain2018) [inline frontend/tests/controllers-test/challengeListCtrl.test.js:3]: Hey @wahid18-maqs , Can you please explain the code you are removing in the tests? and how you are adding the current tests?

File: frontend/src/css/modules/challenge.scss [hunk 4]
```diff
@@ -33,10 +33,10 @@ a.active-challenge {
 }
 
 .ev-challenge-approval-view {
-  margin-top: 0px;
-  padding-top: 30px;
-  padding-bottom: 10px;
-  margin-bottom: 20px;
+    margin-top: 0px;
```
Comment (RishabhJain2018) [inline frontend/src/css/modules/challenge.scss:36]: Why do we have so many unecessary changes here?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Comment by wahid18-maqs** [comment]
### Changes made in Unit tests for challenge list controller

- Added filter panel (organization, tags, start date, end date).
- Implemented search, sort, and `applyFilter()` / `resetFilters()` logic.
- Displayed time in GMT with timezone abbreviation using `moment.tz.guess()`.
- Truncated long descriptions via `isLarge` flag.
- Added scroll-to-top visibility logic.
- Improved error handling, loader states, and digest cycle triggers.
- Wrote comprehensive unit tests for `ChallengeListCtrl`.
- Covered filtering, sorting, and edge cases.
- Mocked dependencies (`$templateCache`, `moment`, `utils`) for isolation.
- Enhanced test coverage and controller reliability.

### Screenshot
eval-filter-test

**Comment by wahid18-maqs** [comment]
@RishabhJain2018 Sir I have made updates to unit testing code and improved CSS styles while following the organization's code structure and conventions.
Review it!!

**Comment by RishabhJain2018** [comment]
Okay, thanks! Hey @akanshajain231999 , Can you please review this?
